### PR TITLE
planner: chose outer table based on cost when both tables are specified in TIDB_INLJ

### PR DESCRIPTION
**Pull request by zz-jason**
### What problem does this PR solve? 

fix https://github.com/pingcap/tidb/issues/9577

### What is changed and how it works?

only short circuit when only one table is specified in the Index Join Hint.

### Check List 

Tests 

 - Integration test
 - Manual test (add detailed scripts or steps below)

Related changes

 - Need to cherry-pick to the release branch
 - Need to be included in the release note

### CODE REVIEW

File: planner/core/exhaust_physical_plans.go [hunk 1]
```diff
@@ -614,7 +614,7 @@ func (p *LogicalJoin) tryToGetIndexJoin(prop *property.PhysicalProperty) ([]Phys
 		join := p.getIndexJoinByOuterIdx(prop, 0)
 		if join != nil {
 			// If the plan is not nil and matches the hint, return it directly.
```
Comment (lamxTyler) [inline planner/core/exhaust_physical_plans.go]: Update comment?

File: planner/core/exhaust_physical_plans.go [hunk 2]
```diff
@@ -598,62 +599,68 @@ func (p *LogicalJoin) buildFakeEqCondsForIndexJoin(keys, idxCols []*expression.C
...
 			return leftJoins, true
 		}
 
 		rightJoins := p.getIndexJoinByOuterIdx(prop, 1)
-		if rightOuter && rightJoins != nil {
+		if rightJoins != nil && rightOuter && !leftOuter {
 			return rightJoins, true
 		}
 
+		hasIndexJoinHint := leftOuter || rightOuter
```
Comment (eurekaka) [inline planner/core/exhaust_physical_plans.go]: Put this line after line 604 so we can use `hasIndexJoinHint` in line 607?

Comment (zz-jason) [reply planner/core/exhaust_physical_plans.go]: done

File: planner/core/exhaust_physical_plans.go [hunk 3]
```diff
@@ -598,62 +599,68 @@ func (p *LogicalJoin) buildFakeEqCondsForIndexJoin(keys, idxCols []*expression.C
...
 		}
 
 		if rightJoins != nil && rhsCardinality < lhsCardinality {
-			return rightJoins, rightOuter
+			return rightJoins, hasIndexJoinHint
 		}
 
-		plans = append(plans, leftJoins...)
-		plans = append(plans, rightJoins...)
+		joins := append(leftJoins, rightJoins...)
```
Comment (winoros) [inline planner/core/exhaust_physical_plans.go]: Is it necessary to go here?
The above two if branches should select the . wanted choice already.

Comment (zz-jason) [reply planner/core/exhaust_physical_plans.go]: if `lhsCardinality == lhsCardinality`, it's possible to reach this place.

**Pull request merged**